### PR TITLE
Make animated image playback speed match browser (Firefox and Chrome) behaviour

### PR DESCRIPTION
**Pull request by goldbattle**
Pull request checklist:

- [x] `CHANGELOG.md` was updated, if applicable

# Description

Animated emotes can have very fast animation compared to the twitch web viewer. This is caused by the browser enforcing a minimum frame time. For example an emote could be animated in chatterino at 1ms per frame, but in any web browser the actual framerate is 100ms for frame.  I think most people would expect emotes the behave like in the browser. See this issue for some background: https://github.com/SevenTV/chatterino7/issues/46

Details on how browsers implement it are here:
https://source.chromium.org/chromium/chromium/src/+/fe1bca993584d616f52d9088d6b1411537441e9d:third_party/blink/renderer/platform/graphics/deferred_image_decoder.cc;l=353-360;bpv=1;bpt=0
https://bugzilla.mozilla.org/show_bug.cgi?id=1511298#c7
```cpp
base::TimeDelta DeferredImageDecoder::FrameDurationAtIndex(size_t index) const {
  base::TimeDelta duration;
  if (metadata_decoder_)
    duration = metadata_decoder_->FrameDurationAtIndex(index);
  if (index < frame_data_.size())
    duration = frame_data_[index].duration_;
  // Many annoying ads specify a 0 duration to make an image flash as quickly as
  // possible. We follow Firefox's behavior and use a duration of 100 ms for any
  // frames that specify a duration of <= 10 ms. See <rdar://problem/7689300>
  // and <http://webkit.org/b/36082> for more information.
  if (duration <= base::TimeDelta::FromMilliseconds(10))
    duration = base::TimeDelta::FromMilliseconds(100);
  return duration;
}
```

### CODE REVIEW

**Comment by Felanbird** [comment]
I don't have this issue so I'm not sure if this fixes it, but both PR's will need a Changelog entry either way.

**Comment by jammehcow** [comment]
[SevenTV#46 is](https://github.com/SevenTV/chatterino7/issues/46) is actually related to missing frames, unrelated :)

File: src/messages/Image.cpp [hunk 1]
```diff
@@ -152,8 +152,15 @@ namespace detail {
             if (reader.read(&image))
             {
                 QPixmap::fromImage(image);
-
-                int duration = std::max(20, reader.nextImageDelay());
+                // It seems that browsers have special logic for fast animations.
+                // This implments Chrome and Firefox's behavior which uses
```
Comment (pajlada) [inline src/messages/Image.cpp:156]:
```suggestion
                // This implements Chrome and Firefox's behavior which uses
```

File: src/messages/Image.cpp [hunk 2]
```diff
@@ -152,8 +152,15 @@ namespace detail {
...
+                // a duration of 100 ms for any frames that specify a duration of <= 10 ms.
+                // See <http://webkit.org/b/36082> for more information.
```
Comment (pajlada) [inline src/messages/Image.cpp:158]:
```suggestion
                // See http://webkit.org/b/36082 for more information.
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -3,6 +3,7 @@
 ## Unversioned
 
 - Major: Added customizable shortcuts. (#2340)
+- Minor: Make animated emote playback speed match browser (#3506)
```
Comment (pajlada) [inline CHANGELOG.md:6]:
```suggestion
- Minor: Make animated emote playback speed match browser (Firefox and Chrome) behaviour. (#3506)
```

**Review by zneix — APPROVED** [review]

**Review by pajlada — CHANGES_REQUESTED** [review]

**Comment by pajlada** [comment]
Thanks for your contribution! As a first-time contributor, you can now add yourself to the contributors list that's shown inside the application.

To do so, open a new PR where you modify the `resources/contributors.txt` file and add yourself to the list. Make sure to read the comments at a top for instructions.

**Pull request merged**